pkg/network/ovn_kubernetes.go

```diff
@@ -1840,3 +1731,196 @@ func isIPSecEnabledInPod(pod v1.PodTemplateSpec, containerName string) bool {
+			continue
+		}
+		if utilnet.IsIPv6CIDRString(cn.CIDR) {
+			cnHasIPv6 = true
+		} else {
+			cnHasIPv4 = true
+		}
+		if err := pool.Add(*cidr); err != nil {
+			out = append(out, errors.Errorf("Whole or subset of ClusterNetwork CIDR %s is already in use: %s", cn.CIDR, err))
+		}
+	}
+	for _, snet := range conf.ServiceNetwork {
+		_, cidr, err := net.ParseCIDR(snet)
+		if err != nil {
+			out = append(out, errors.Wrapf(err, "could not parse spec.serviceNetwork %s", snet))
+			continue
+		}
+		if err := pool.Add(*cidr); err != nil {
+			out = append(out, errors.Errorf("Whole or subset of ServiceNetwork CIDR %s is already in use: %s", snet, err))
+		}
+	}
+
+	oc := conf.DefaultNetwork.OVNKubernetesConfig
+	// Note: oc.V4InternalSubnet will be deprecated in future as per k8s guidelines
+	// oc.V4InternalSubnet and oc.IPv4.InternalJoinSubnet must be same if both are present
+	if oc.V4InternalSubnet != "" && oc.IPv4 != nil {
+		if oc.IPv4.InternalJoinSubnet != "" && oc.V4InternalSubnet != oc.IPv4.InternalJoinSubnet {
+			out = append(out, errors.Errorf("v4InternalSubnet will be deprecated soon, until then it must be same as v4InternalJoinSubnet %s ", oc.IPv4.InternalJoinSubnet))
+
+		} else if oc.IPv4.InternalJoinSubnet != "" && oc.V4InternalSubnet == oc.IPv4.InternalJoinSubnet {
+			if !cnHasIPv4 {
+				out = append(out, errors.Errorf("JoinSubnet %s and ClusterNetwork must have matching IP families", oc.IPv4.InternalJoinSubnet))
+			}
+			// If both InternalSubnet and InternalJoinSubnet has been added then CNO should not exhibit
+			// an overlap error. So, avoiding addition of InternalSubnet to IPPpool
+			if err := validateOVNKubernetesSubnet("v4InternalJoinSubnet", oc.IPv4.InternalJoinSubnet, &pool, conf.ClusterNetwork); err != nil {
+				out = append(out, err)
+			}
+		} else {
+			if !cnHasIPv4 {
+				out = append(out, errors.Errorf("JoinSubnet %s and ClusterNetwork must have matching IP families", oc.V4InternalSubnet))
+			}
+			if err := validateOVNKubernetesSubnet("v4InternalSubnet", oc.V4InternalSubnet, &pool, conf.ClusterNetwork); err != nil {
+				out = append(out, err)
+			}
+		}
+	} else if oc.V4InternalSubnet != "" && oc.IPv4 == nil {
+		if !cnHasIPv4 {
+			out = append(out, errors.Errorf("JoinSubnet %s and ClusterNetwork must have matching IP families", oc.V4InternalSubnet))
+		}
+		if err := validateOVNKubernetesSubnet("v4InternalSubnet", oc.V4InternalSubnet, &pool, conf.ClusterNetwork); err != nil {
+			out = append(out, err)
+		}
+	} else if oc.V4InternalSubnet == "" && oc.IPv4 != nil {
+		if oc.IPv4.InternalJoinSubnet != "" {
+			if !cnHasIPv4 {
+				out = append(out, errors.Errorf("JoinSubnet %s and ClusterNetwork must have matching IP families", oc.IPv4.InternalJoinSubnet))
+			}
+			if err := validateOVNKubernetesSubnet("v4InternalJoinSubnet", oc.IPv4.InternalJoinSubnet, &pool, conf.ClusterNetwork); err != nil {
+				out = append(out, err)
+			}
+		}
+	}
```

Comment:
We can go for a shorter approach if we use an aux variable to store the subnet
```
v4InternalSubnet := oc.V4InternalSubnet
if oc.IPv4 != nil && oc.IPv4.InternalJoinSubnet != "" {
    if v4InternalSubnet != "" && v4InternalSubnet != oc.IPv4.InternalJoinSubnet {
        out = append(out, errors.Errorf("v4InternalSubnet will be deprecated soon, until then it must be same as v4InternalJoinSubnet %s ", oc.IPv4.InternalJoinSubnet))
    }
    v4InternalSubnet = oc.IPv4.InternalJoinSubnet
}
if v4InternalSubnet != "" {
    if !cnHasIPv4 {
        out = append(out, errors.Errorf("JoinSubnet %s and ClusterNetwork must have matching IP families", oc.IPv4.InternalJoinSubnet))
    }
    if err := validateOVNKubernetesSubnet("v4InternalJoinSubnet", v4InternalSubnet, &pool, conf.ClusterNetwork); err != nil {
        out = append(out, err)
    }
}
```

And looking at the section of code above, it doesn't look like we are validating if a v4InternalSubnet is present when it should, as if we should do something else with the cnHasIPv4 check:
```
v4InternalSubnet := oc.V4InternalSubnet
if oc.IPv4 != nil && oc.IPv4.InternalJoinSubnet != "" {
    if v4InternalSubnet != "" && v4InternalSubnet != oc.IPv4.InternalJoinSubnet {
        out = append(out, errors.Errorf("v4InternalSubnet will be deprecated soon, until then it must be same as v4InternalJoinSubnet %s ", oc.IPv4.InternalJoinSubnet))
    }
    v4InternalSubnet = oc.IPv4.InternalJoinSubnet
}
if cnHasIPv4 && v4InternalSubnet == "" {
    out = append(out, errors.Errorf("JoinSubnet %s and ClusterNetwork must have matching IP families", oc.IPv4.InternalJoinSubnet))
}
if v4InternalSubnet != "" {
    if err := validateOVNKubernetesSubnet("v4InternalJoinSubnet", v4InternalSubnet, &pool, conf.ClusterNetwork); err != nil {
        out = append(out, err)
    }
}
```

Same approach on the other instances.

Reply:
> Thank you so much @jcaamano for this really nice suggestion. first code block seems sufficient to me.
> 
> ```
> if cnHasIPv4 && v4InternalSubnet == "" {
>     out = append(out, errors.Errorf("JoinSubnet %s and ClusterNetwork must have matching IP families", oc.IPv4.InternalJoinSubnet))
> }
> ```
> 
> Neither oc.V4InternalSubnet nor oc.IPv4.InternalJoinSubnet is a mandatory field and can be empty. So above check seems not required to me. What do you think?

Okey. I got that wrong and I guess we are assuming defaults on ovn-k.

Reply:
yes, right